Jump to next / previous chapter

lks-nbg opened the pull request:
I worked on the second part of issue #6 and tried to implement jumping to next and previous chapters from the player and the notification. Jumping to the next or previous chapter basically works but I'm struggling on how to update currentlyPlaying information. Therefore the actions currently only work once, since the progress is not updated :-/ 

Furthermore I had to adjust the player ui to accommodate two more buttons. I added another row which has a lot of free space at the moment, but it can be used for new features like bookmarks (#11 ). 

I would really appreciate your review and hope to get some tips to finish this pr.

Resolves #6

Commit:
Skip to next/previous chapter

Commit:
Skip to next/previous chapter

lks-nbg commented:
Closes #6

Commit:
Added sort by year option

mattttvaughn commented:
I dig the updated design! 

I don't have a fully-formed idea for how to do all of this, but take this with a grain of salt, but here are some ideas which might help:

For your chapterIndex issue where Plex is reporting 1-based indexing, you can use `currentBook.chapters.indexOf(currentChapter)` to get a consistent 0-based index. 

As far as updating CurrentlyPlaying goes, I'm thinking you _should_ be able to do this via a manual call to `ProgressUpdater.updateProgress` wherever needed since that will update currentlyPlaying, but I haven't thought through the full details. You could even use a convenience like below for `updateProgress()` so you don't have to provide all of those parameters everywhere you call it

```
    override fun updateProgress() {
        val controller = mediaController ?: return
        val playbackState = when (controller.playbackState.state) {
            PlaybackStateCompat.STATE_PLAYING -> MediaPlayerService.PLEX_STATE_PLAYING
            PlaybackStateCompat.STATE_PAUSED -> MediaPlayerService.PLEX_STATE_PAUSED
            PlaybackStateCompat.STATE_STOPPED -> MediaPlayerService.PLEX_STATE_PAUSED
            else -> ""
        }
        val currentTrack = controller.metadata.id?.toInt() ?: return
        val currentTrackProgress = controller.playbackState.currentPlayBackPosition
        updateProgress(
           currentTrack,
            playbackState,
            currentTrackProgress,
            false
        )
    }
```

mattttvaughn commented on app/src/main/java/io/github/mattpvaughn/chronicle/features/player/AudiobookMediaSessionCallback.kt:
```diff
@@ -144,11 +175,19 @@ class AudiobookMediaSessionCallback @Inject constructor(
             // pause/play for inline wired headphones
             return when (ke.keyCode) {
-                KEYCODE_MEDIA_NEXT, KEYCODE_MEDIA_SKIP_FORWARD -> {
+                KEYCODE_MEDIA_NEXT -> {
```
excellent idea 👍

Commit:
Added sort by year option

Commit:
Skip to next/previous chapter

lks-nbg commented:
I fixed the index and moved most of the skipToNext & skipToPrevious code to PlayerExt.kt. Your updateProgress idea is working and you can now skip between chapters as expected. 

There is one issue left with the notification. After skipping to the another chapter (in player view or notification) the notification get's updated but shows the previous chapter title until it's rebuild again.

Resolves #6

mattttvaughn committed:
Merge remote-tracking branch 'origin/sort-by-year' into feature/skip-to-next-previous

mattttvaughn committed:
Merge remote-tracking branch 'origin/feature/skip-to-next-previous' into feature/skip-to-next-previous

mattttvaughn commented on app/src/main/java/io/github/mattpvaughn/chronicle/features/player/PlayerExt.kt:
```diff
@@ -24,3 +30,38 @@ fun Player.seekRelative(trackListStateManager: TrackListStateManager, durationMi
+    if(nextChapterIndex < currentlyPlaying.book.value.chapters.size) {
+        val nextChapter = currentlyPlaying.book.value.chapters[nextChapterIndex]
+        Timber.d("NEXT CHAPTER: index=$nextChapterIndex id=${nextChapter.id} trackId=${nextChapter.trackId}  offset=${nextChapter.startTimeOffset} title=${nextChapter.title}")
+        seekTo(trackListStateManager.currentTrackIndex, nextChapter.startTimeOffset)
```
in a book where each chapter is its own track, this would seek to the start of the current track (since the next chapter will have an offset of 0 since it starts at 0 of it's own track). You can get the index of the chapter's track with something like this:

```
    val containingTrack = trackListStateManager.trackList
        .firstOrNull {
            it.id.toLong() == previousChapter.trackId
        }
    val containingTrackIndex = trackListStateManager.trackList.indexOf(containingTrack)

```

mattttvaughn commented on app/src/main/java/io/github/mattpvaughn/chronicle/features/player/PlayerExt.kt:
```diff
@@ -24,3 +30,38 @@ fun Player.seekRelative(trackListStateManager: TrackListStateManager, durationMi
+    if(previousChapterIndex < 0) previousChapterIndex = 0
+    val previousChapter = currentlyPlaying.book.value.chapters[previousChapterIndex]
+    Timber.d("PREVIOUS CHAPTER: index=$previousChapterIndex id=${previousChapter.id} trackId=${previousChapter.trackId}  offset=${previousChapter.startTimeOffset} title=${previousChapter.title}")
+    seekTo(trackListStateManager.currentTrackIndex, previousChapter.startTimeOffset)
```
ditto above

mattttvaughn commented on app/src/main/java/io/github/mattpvaughn/chronicle/features/currentlyplaying/CurrentlyPlayingViewModel.kt:
```diff
@@ -372,31 +373,30 @@ class CurrentlyPlayingViewModel(
                 Timber.i("Seeking!")
                 transportControls?.sendCustomAction(action, null)
             } else {
+                val currentChapterIndex = currentlyPlaying.book.value.chapters.indexOf(currentlyPlaying.chapter.value)
```
looks good 👍

mattttvaughn reviewed (commented):
left the one comment about multi-track books, but looking good!

One other thing I'm seeing- pulled this down and am seeing an issue with skipping forwards while a book is paused. After skipping forwards, seeing progress as being at the end of the current chapter, and not in the next chapter we are skipping to (for example, progress will show as 23:30/23:30).

This isn't a big issue, except that it prevents from skipping multiple chapters at a time cuz they get stuck at the end of a chapter and keep skipping to where they're at.

At a glance (and just one my one example book it tried) it can be solved by changing positionMs parameter in `Player.skipToNext()` to be nextChapter.startTimeOffset + 1, but might be worth digging into a little.

mattttvaughn commented:
and @lks-nbg on the notification issue. I'd think calling NotificationBuilder would be the best way to go about this

Providing the following to the skipTo functions should do it:

```
    notificationBuilder: NotificationBuilder,
    controller: MediaControllerCompat,
    serviceScope: CoroutineScope
) {

```

and then at the end of `skipToPrev/Next()`

```
        serviceScope.launch {
            notificationBuilder.buildNotification(controller.sessionToken)
        }

```

And then just providing via `@Inject`s and parent functions where needed.

It's gonna be ugly, but we can eventually just build a `Player` wrapper at some point to replace `PlayerExt`

Commit:
Skip to next/previous chapter

lks-nbg commented:
Thanks for your comments. I tried to implement them:
- skipToNext + 300ms (like in AudiobookMediaSessionCallback) 
- multi-track → I added your code but since I only have single-track audiobooks I cannot test it 
- NotificationBuilder → not sure if I missed something, but my notifications are still one chapter behind

mattttvaughn committed:
Merge remote-tracking branch 'origin/feature/skip-to-next-previous' into feature/skip-to-next-previous

mattttvaughn committed:
Fix notification not updating on chapter skip

mattttvaughn commented:
@lks-nbg Added a commit to fix that notification not building issue, that was my fault, the `val notification = notificationBuilder.buildNotification(...)` doesn't actually set the notification, it just creates it.

Give it a whirl and lmk if you run into any issues. Looking pretty good from my end

Commit:
Simplified notification update code after chapter skip

lks-nbg commented:
I think I found a more elegant solution, where we can get rid of the ugly code in PlayerExt. The callback onChapterChange is already implemented but was not used so far. 

What do you think about this?

mattttvaughn reviewed (approved):
looks solid

mattttvaughn commented:
Good idea, looks good to me.

At this point I'm happy to merge whenever you're happy with it, just lmk. Can put out an unofficial build and get some people to try it out against edge cases.

Also, maybe worth noting that it looks like there is an unrelated bug in `OnMediaChangedCallback` which is causing multi-track books not to update chapter names in the notification on chapter change, but it works on my end after fixing that. I'll push a separate commit to fix that after merging this.

lks-nbg commented:
For now I am happy with the code and you can merge it. 
Getting some feedback is a good idea 👍
